File: java/src/sync/FileInfo.java
package sync;

import java.util.List;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class FileInfo {
  public Path name; // a relative path, relative to the base directory
  public FileHash hash;
  public long size;

  public FileInfo(Path name, FileHash hash, long size) {
    this.name = name;
    this.hash = hash;
    this.size = size;
  }
  public FileInfo(Path path, Path base) throws IOException {
    this.name = base.relativize(path);
    this.size = Files.size(path);
    this.hash = new FileHash(path);
  }
  // Reads the directory and returns a list of all files in the directory
  static List<FileInfo> readList(Path dir) throws IOException {
    try (var s = Files.find(dir, Integer.MAX_VALUE, (p, a) -> a.isRegularFile())) {
      return s.map(p -> {
        try {
          return new FileInfo(p, dir);
        } catch (IOException e) {
          throw new RuntimeException(e); // this shouldn't happen, since the file must exist to get here
        }
      }).toList();
    }
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    FileInfo fileInfo = (FileInfo) o;
    return name.equals(fileInfo.name) && hash.equals(fileInfo.hash) && size == fileInfo.size;
  }
};

File: ts/src/common/FileInfo.ts
import fs from "fs/promises";
import path from "path";
import { FileHash } from "./FileHash.ts";

export class FileInfo {
  name: string; // a relative path, relative to the base directory
  hash: FileHash;
  size: number;

  constructor(name: string, hash: FileHash, size: number) {
    this.name = name;
    this.hash = hash;
    this.size = size;
  }
  static async fromPath(file: string, base: string): Promise<FileInfo> {
    const name = path.relative(base, file);
    const [hash, size] = await Promise.all([
      FileHash.fromPath(file),
      fs.stat(file).then(s => s.size),
    ]);
    return new FileInfo(name, hash, size);
  }
  // Reads the directory and returns a list of all files in the directory
  static async readList(dirpath: string): Promise<FileInfo[]> {
    const files: Promise<FileInfo>[] = [];
    for await (const dirent of await fs.opendir(dirpath, { recursive: true })) {
      if (!dirent.isFile()) continue;
      const name = path.join(dirpath, dirent.name);
      files.push(FileInfo.fromPath(name, dirpath));
    }
    return Promise.all(files);
  }
}
